lib/diagnostics_channel.js

```diff
@@ -0,0 +1,143 @@
+'use strict';
+
+const {
+  ArrayPrototypeIncludes,
+  ArrayPrototypePush,
+  FunctionPrototypeBind,
+  ObjectCreate,
+  ObjectSetPrototypeOf,
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+  }
+} = require('internal/errors');
+
+// TODO(qard): should there be a C++ publisher interface?
+class ActivePublisher {
+  subscribe(subscription) {
+    if (typeof subscription === 'function') {
+      ArrayPrototypePush(this._subscribers, subscription);
+      return;
+    }
+
+    if (typeof subscription === 'object' &&
+        typeof subscription.onMessage === 'function') {
+      let shouldListen = true;
+
+      if (typeof subscription.shouldListen === 'function') {
+        shouldListen = subscription.shouldListen(this.name);
+      }
+
+      if (shouldListen) {
+        const fn = FunctionPrototypeBind(subscription.onMessage, subscription);
+        ArrayPrototypePush(this._subscribers, fn);
+      }
+    }
+  }
+
+  shouldPublish() {
+    return true;
+  }
+
+  publish(data) {
+    for (var i = 0; i < this._subscribers.length; i++) {
+      this._subscribers[i](this.name, data);
+    }
+  }
+}
+
+class Publisher {
+  constructor(name) {
+    this._subscribers = undefined;
+    this.ifActive = undefined;
+    this.name = name;
+  }
+
+  subscribe(subscription) {
+    ObjectSetPrototypeOf(this, ActivePublisher.prototype);
+    this.ifActive = this;
+    this._subscribers = [];
+    this.subscribe(subscription);
+  }
+
+  shouldPublish() {
+    return false;
+  }
+
+  publish() {}
+}
+
+class DiagnosticsChannel {
+  constructor() {
+    this._channels = ObjectCreate(null);
+    this._subscribers = [];
+  }
+
+  publisher(name) {
+    let channel = this._channels[name];
+    if (channel) return channel;
+
+    if (typeof name !== 'string') {
+      throw new ERR_INVALID_ARG_TYPE('publisher', ['string'], name);
+    }
+
+    channel = new Publisher(name);
+
+    // New channels should check existing subscribers
+    for (let i = 0; i < this._subscribers.length; i++) {
+      const subscriber = this._subscribers[i];
+      channel.subscribe(subscriber);
+    }
+
+    this._channels[name] = channel;
```

Comment:
This creates a memory leak if `name` includes some dynamic parts.

Reply:
I'm not sure what you mean. Dynamic parts? It should only be possible to reach that point if it's a plain string.

Reply:
Oh, you mean like having a user id in the channel name or something like that? That's not how this is meant to be used. The intent is that a module would create one or maybe a few publisher objects at the top-level of the module and just reuse those. Channels should probably not be dynamically generated. It's not communicated well in the docs yet, those were somewhat hastily thrown together and need a lot more detail. 😅 

Imagine something like this:

```js
const dc = require('diagnostics_channel')
const calls = dc.publisher('my-function-calls')

let count = 0
module.exports = function myFunction(a, b, callback) {
  const id = ++count
  calls.ifActive?.publish({ span: 'start', id, a, b })
  setImmediate(() => {
    calls.ifActive?.publish({ span: 'end', id, a, b })
    callback(a, b)
  })
}

// elsewhere

dc.subscribe('my-function-calls', (name, data) => {
  // capture start and end events to create a span record
})
```

Reply:
I was also considering the idea of giving every module file a unique channel object named after its module path, probably under a getter so it's only created if it is used. For example `module.channel.ifActive?.publish(...)` used in mongodb might be named `mongodb/index.js`.

Reply:
I'm wary of the per module channel, especially since you can synthesize modules and ESM is not based upon the Module prototype.

Reply:
I'm definitely not tied to that idea. Just something I thought of that might be more easily controlled than just handing the keys over to the user to create channels wherever and whenever they want. ESM would probably need its own thing if we wanted to do anything with that idea, though we may just continue with the current path of just leaving it up to the user to explicitly construct their own channels.

Reply:
There should be some logic that prevents/warns in case of a memory leak situation. I think it's also possible to avoid the leak altogether if the publisher goes out of scope (weakrefs).

Reply:
I don't know that I'd characterize it as a leak, but I agree that it'd be good to at least have a warning if there's more than n channels created. I'll experiment a bit with the weakref idea. Being able to let channels go out of scope and get gc'd would be nice just in case people try to use them more dynamically, though I wouldn't really expect _any_ to go out of scope in the _suggested_ use. As I said, they _should_ be more of top-level module scope thing, not dynamically generated.

Reply:
I think using a `WeakRef` for this would solve this problem.